hudi-aws/src/test/java/org/apache/hudi/aws/transaction/lock/TestS3StorageLockClient.java

```diff
@@ -0,0 +1,380 @@
+            mockLogger
+    );
+  }
+
+  private void mockS3ObjectWithLockData(StorageLockData lockData, String eTag) {
+    byte[] bytes = StorageLockFile.toByteArray(lockData);
+    ByteArrayInputStream bais = new ByteArrayInputStream(bytes);
+
+    ResponseInputStream<GetObjectResponse> mockResponseStream = new ResponseInputStream<>(
+            GetObjectResponse.builder().eTag(eTag).build(),
+            new MockAbortableInputStream(bais)
+    );
+
+    when(mockS3Client.getObject(any(GetObjectRequest.class))).thenReturn(mockResponseStream);
+  }
+
+  private static class MockAbortableInputStream extends java.io.FilterInputStream {
+    protected MockAbortableInputStream(ByteArrayInputStream in) {
+      super(in);
+    }
+  }
+
+  @Test
+  void testTryCreateOrUpdateLockFile_noPreviousLock_success() {
+    StorageLockData lockData = new StorageLockData(false, System.currentTimeMillis(), "myTxOwner");
+    PutObjectResponse putResp = PutObjectResponse.builder().eTag("new-etag-123").build();
+    when(mockS3Client.putObject(any(PutObjectRequest.class), any(RequestBody.class))).thenReturn(putResp);
+
+    Pair<LockUpdateResult, StorageLockFile> result = lockService.tryCreateOrUpdateLockFile(lockData, null);
+
+    assertEquals(LockUpdateResult.SUCCESS, result.getLeft());
+    assertNotNull(result.getRight());
+    assertEquals("new-etag-123", result.getRight().getVersionId());
+    verify(mockS3Client, times(1)).putObject(any(PutObjectRequest.class), any(RequestBody.class));
+    verifyNoMoreInteractions(mockLogger);
+  }
+
+  @Test
+  void testTryCreateOrUpdateLockFile_withPreviousLock_success() {
+    StorageLockData lockData = new StorageLockData(false, 1000L, "myTxOwner");
+    StorageLockFile prevLockFile = new StorageLockFile(lockData, "old-etag-999");
+    PutObjectResponse putResp = PutObjectResponse.builder().eTag("new-etag-456").build();
+    when(mockS3Client.putObject(
+            eq(PutObjectRequest.builder().bucket(BUCKET_NAME)
+                    .key(LOCK_FILE_PATH)
+                    .ifMatch("old-etag-999")
+                    .build()),
+            any(RequestBody.class)
+    )).thenReturn(putResp);
+
+    Pair<LockUpdateResult, StorageLockFile> result =
+            lockService.tryCreateOrUpdateLockFile(lockData, prevLockFile);
+
+    assertEquals(LockUpdateResult.SUCCESS, result.getLeft());
+    assertNotNull(result.getRight());
+    assertEquals("new-etag-456", result.getRight().getVersionId());
+    verify(mockS3Client, times(1)).putObject(any(PutObjectRequest.class), any(RequestBody.class));
+  }
+
+  @Test
+  void testTryCreateOrUpdateLockFile_preconditionFailed() {
+    StorageLockData lockData = new StorageLockData(false, 2000L, "txOwner");
+    StorageLockFile prevLockFile = new StorageLockFile(lockData, "some-etag");
+
+    AwsServiceException ex412 = S3Exception.builder().statusCode(412).build();
+    when(mockS3Client.putObject(any(PutObjectRequest.class), any(RequestBody.class))).thenThrow(ex412);
```

Comment:
so its mismatching due to the mocked etag. ?

Reply:
Yes